delira/utils/config.py

```diff
@@ -1,13 +1,509 @@
+        else:
+            # set item
+            self._set_internal_item(key, item, deepcopy=deepcopy)
+
+    def _raise_overwrite(self, key, overwrite):
+        """
+        Checks if a ValueError should be raised
+
+        Parameters
+        ----------
+        key : str
+            key which needs to be checked
+        overwrite : bool
+            if overwrite ir enables no error is raised even if key already
+            exists
+
+        Raises
+        ------
+        ValueError
+            if overwrite is not enabled and key already exists
+        """
+        if key in self and not overwrite:
+            raise ValueError("{} already in config. Can "
+                             "not overwrite value.".format(key))
+
+    def _set_internal_item(self, key, item, deepcopy=False):
+        """
+        Set internal item
+
+        Parameters
+        ----------
+        key : str
+            key where new item should be assigned
+        item : Any
+            item which should be assigned
+        deepcopy : bool, optional
+            if enabled the item is copied to the config, by default False
+        """
+        config_item = self._to_config(item)
+        if deepcopy:
+            self[key] = copy.deepcopy(config_item)
+        else:
+            self[key] = config_item
+
+    @staticmethod
+    def _create_internal_dict(*args, **kwargs):
+        """
+        Defines how internal dicts should be created. Can be used to easily
+        overwrite subclasses
+
+        Returns
+        -------
+        :class:`Config`
+            new config
+        """
+        return Config(*args, **kwargs)
+
+    @classmethod
+    def _to_config(cls, item):
+        """
+        Convert items to config if they are a dict like object
+        but not already a config
+
+        Parameters
+        ----------
+        item : Any
+            item which is converted
+
+        Returns
+        -------
+        Any
+            return a config is item is dict like, otherwise the item is
+            returned
+        """
+        if isinstance(item, dict) and not isinstance(item, cls):
+            # convert dict to config for additional functionality
+            return cls._create_internal_dict(item)
+        else:
+            return item
+
+    def dump(self, path, formatter=yaml.dump, encoder_cls=Encoder, **kwargs):
+        """
+        Save config to a file and add time stamp to config
+
+        Parameters
+        ----------
+        path : str
+            path where config is saved
+        formatter : callable, optional
+            defines the format how the config is saved, by default yaml.dump
+        encoder_cls : :class:`Encoder`, optional
+            trasforms config to a format which can be formatted by the
+            `formatter`, by default Encoder
+        kwargs:
+            additional keyword arguments passed to `formatter`
+        """
+        self._timestamp = now()
```

Comment:
Maybe we should initialize this with None and only change the timestamp if it hasn't been set to anything different so far?

Reply:
We can do this. I removed it from the `__init__` because it would overwrite the original timestamp if the config was loaded.

Reply:
But in the `__init__` you would first initialize it with `None` and then overwrite it with the loaded values

Reply:
I tried adding it to the `__init__` and the result was too messy because every internal dictionary is a config itself and thus has its own timestamp. Do we want a separate argument to add the timestamp to the config at initialisation or do we leave it like the current implementation (added during dumping the config)?

Reply:
No, we leave it as is. It was just a suggestion.